worker/imgproxy.js

```diff
@@ -0,0 +1,194 @@
+  u => u.host === 'stacker.news',
+  u => u.host === 'news.ycombinator.com',
+  u => u.host === 'www.youtube.com' || u.host === 'youtu.be',
+  u => u.host === 'github.com'
+]
+
+function matchUrl(matchers, url) {
+  try {
+    return matchers.some(matcher => matcher(new URL(url)))
+  } catch(err) {
+    console.log(url, err)
+    return false
+  }
+}
+
+function decodeOriginalUrl (imgproxyUrl) {
+  const parts = imgproxyUrl.split('/')
+  const b64Url = parts[parts.length - 1]
+  const originalUrl = Buffer.from(b64Url, 'base64url').toString('utf-8')
+  return originalUrl
+}
+
+export function imgproxy ({ models }) {
+  return async function ({ data: { id, forceFetch = false} }) {
+    if (!imgProxyEnabled) return
+
+    console.log('running imgproxy job', id)
+
+    const item = await models.item.findUnique({ where: { id } })
+
+    const isJob = typeof item.maxBid !== 'undefined'
+
+    let imgproxyUrls = {}
+    try {
+      if (item.text) {
+        imgproxyUrls = await createImgproxyUrls(id, item.text, { forceFetch })
+      }
+      if (item.url && !isJob) {
+        imgproxyUrls = { ...imgproxyUrls, ...(await createImgproxyUrls(id, item.url, { forceFetch })) }
+      }
+    } catch(err) {
+      console.log("[imgproxy] error:", err)
+      // rethrow for retry
+      throw err
+    }
+
+    console.log("[imgproxy] updating item", id, "with urls", imgproxyUrls)
+
+    await models.item.update({ where: { id }, data: { imgproxyUrls } })
+  }
+}
+
+export const createImgproxyUrls = async (id, text, { forceFetch }) => {
+  const urls = extractUrls(text)
+  console.log("[imgproxy] id:", id, "-- extracted urls:", urls)
+  // resolutions that we target:
+  //   - nHD:  640x 360
+  //   - qHD:  960x 540
+  //   - HD:  1280x 720
+  //   - HD+: 1600x 900
+  //   - FHD: 1920x1080
+  //   - QHD: 2560x1440
+  // reference:
+  //   - https://en.wikipedia.org/wiki/Graphics_display_resolution#High-definition_(HD_and_derivatives)
+  //   - https://www.browserstack.com/guide/ideal-screen-sizes-for-responsive-design
+  const resolutions = ['640x360', '960x540', '1280x720', '1600x900', '1920x1080', '2560x1440']
+  const imgproxyUrls = {}
+  for (let url of urls) {
+    if(!url) continue
+
+    console.log("[imgproxy] id:", id, "-- processing url:", url)
+    if (url.startsWith(IMGPROXY_URL)) {
+      console.log("[imgproxy] id:", id, "-- proxy url, decoding original url:", url)
+      // backwards compatibility: we used to replace image urls with imgproxy urls
+      url = decodeOriginalUrl(url)
+      console.log("[imgproxy] id:", id, "-- original url:", url)
+    }
+    if (!(await isImageURL(url, { forceFetch }))) {
+      console.log("[imgproxy] id:", id, "-- not image url:", url)
+      continue
+    }
+    imgproxyUrls[url] = {}
+    for (const res of resolutions) {
+      const [w, h] = res.split('x')
+      const processingOptions = `/rs:fill:${w}:${h}`
+      imgproxyUrls[url][`${w}w`] = createImgproxyUrl(url, processingOptions)
```

Comment:
Does this also store the image width and height? 

Currently because we don't know the width and height, we can't reserve space for loading images so when they do load, they go from 0 height to whatever height they are, causing significant layout shift. It'd be nice to remove that.

Reply:
> Does this also store the original image width and height?

Not the original, this `w` here is only the upper bound for the width of the processed image.

But with the `w`, we also know the upper bound for height. So we could at least minimize layout shift by using heuristics?

If we want to know the exact width and height, we need to know the original width and height. From that, we should be able to tell which width and height the processed image in the various resolutions will have.

But for this, we either need to roll our own code or pay for imgproxy since the image info feature is a Pro feature:

> imgproxy can fetch and return a source image info without downloading the whole image.

-- https://docs.imgproxy.net/getting_the_image_info